core/trino-main/src/main/java/io/trino/sql/planner/iterative/rule/PreferWritePartitioning.java

```diff
@@ -0,0 +1,59 @@
+package io.trino.sql.planner.iterative.rule;
+
+import io.trino.Session;
+import io.trino.matching.Captures;
+import io.trino.matching.Pattern;
+import io.trino.sql.planner.iterative.Rule;
+import io.trino.sql.planner.plan.TableWriterNode;
+
+import java.util.Optional;
+
+import static io.trino.SystemSessionProperties.getPreferredWritePartitioningThreshold;
+import static io.trino.SystemSessionProperties.isUsePreferredWritePartitioning;
+import static io.trino.sql.planner.plan.Patterns.tableWriterNode;
+
+public class PreferWritePartitioning
+        implements Rule<TableWriterNode>
+{
+    @Override
+    public Pattern<TableWriterNode> getPattern()
+    {
+        return tableWriterNode();
+    }
+
+    @Override
+    public boolean isEnabled(Session session)
+    {
+        return isUsePreferredWritePartitioning(session);
+    }
+
+    @Override
+    public Result apply(TableWriterNode node, Captures captures, Context context)
+    {
+        if (node.getPreferredPartitioningScheme().isEmpty()) {
+            return Result.empty();
+        }
+
+        double numberOfPartitions = node.getPreferredPartitioningScheme().get().getPartitioning().getColumns().stream()
```

Comment:
Calculating number of partition is non-trivial for multiple partitions.
It is somewhere between max cardinality of one of partitions and product of cardinality of all of them.